File: backend/app/routers/basic_info.py
import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..auth import get_current_user
from ..database import get_db
from ..models import User
from ..repositories import BasicInfoRepository
from ..schemas import BasicInfoCreate, BasicInfoResponse, BasicInfoUpdate

router = APIRouter(prefix="/api/basic-info", tags=["basic-info"])


@router.post("", response_model=BasicInfoResponse, status_code=201)
def create_basic_info(
    payload: BasicInfoCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> BasicInfoResponse:
    repository = BasicInfoRepository(db, current_user.id)
    return repository.create(payload.model_dump())


@router.get("/latest", response_model=BasicInfoResponse)
def get_latest_basic_info(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> BasicInfoResponse:
    repository = BasicInfoRepository(db, current_user.id)
    basic_info = repository.get_latest()
    if not basic_info:
        raise HTTPException(status_code=404, detail="Basic info not found")
    return basic_info


@router.put("/{basic_info_id}", response_model=BasicInfoResponse)
def update_basic_info(
    basic_info_id: uuid.UUID,
    payload: BasicInfoUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> BasicInfoResponse:
    repository = BasicInfoRepository(db, current_user.id)
    basic_info = repository.get_by_id(str(basic_info_id))
    if not basic_info:
        raise HTTPException(status_code=404, detail="Basic info not found")

    return repository.update(basic_info, payload.model_dump())

File: frontend/src/api/basic-info.ts
import type { BasicInfoPayload, BasicInfoResponse } from "../types";
import { request } from "./client";

export const BASIC_INFO_REQUIRED_MESSAGE = "基本情報の氏名と記載日を先に登録してください。";

export function createBasicInfo(payload: BasicInfoPayload): Promise<BasicInfoResponse> {
  return request<BasicInfoResponse>("/api/basic-info", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateBasicInfo(id: string, payload: BasicInfoPayload): Promise<BasicInfoResponse> {
  return request<BasicInfoResponse>(`/api/basic-info/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function getLatestBasicInfo(): Promise<BasicInfoResponse> {
  return request<BasicInfoResponse>("/api/basic-info/latest");
}

export async function assertBasicInfoReady(): Promise<void> {
  try {
    const basicInfo = await getLatestBasicInfo();
    if (!basicInfo.full_name || !basicInfo.record_date) {
      throw new Error(BASIC_INFO_REQUIRED_MESSAGE);
    }
  } catch (error) {
    if (error instanceof Error && error.message === BASIC_INFO_REQUIRED_MESSAGE) {
      throw error;
    }
    throw new Error(BASIC_INFO_REQUIRED_MESSAGE);
  }
}
